src/states/node.rs

```diff
@@ -1877,98 +1852,117 @@ impl Node {
         let dst = Authority::Section(relocation_dst);
         self.send_routing_message(src, dst, request_content)
     }
 
     // Received by Y; From X -> Y
-    // Context: a node is joining our section. Sends `AcceptAsCandidate` to our section. If the
-    // network is unbalanced, sends `ExpectCandidate` on to a section with a shorter prefix.
+    // Context: a node is joining our section. Vote `ExpectCandidate`.
     fn handle_expect_candidate(
         &mut self,
-        old_pub_id: PublicId,
-        old_client_auth: Authority<XorName>,
-        relocation_dst: Authority<XorName>,
-        message_id: MessageId,
+        rpc: ExpectCandidate,
+        dst_name: XorName,
     ) -> Result<(), RoutingError> {
-        self.remove_expired_peers();
+        if !self.chain.is_member() || !self.peer_mgr.is_established() {
+            return Ok(());
+        }
 
-        if old_pub_id == *self.full_id.public_id() {
-            return Ok(()); // This is a delayed message belonging to our own relocate request.
+        self.vote_for_event(ExpectCandidateVote { rpc, dst_name }.into_network_event());
+        Ok(())
+    }
+
+    // Handles an accumulated `ExpectCandidate` event.
+    // Context: a node is joining our section. Sends the node our section. If the
+    // network is unbalanced, sends `ExpectCandidate` on to a section with a shorter prefix.
+    fn handle_expect_candidate_event(
+        &mut self,
+        vote: ExpectCandidateVote,
+    ) -> Result<(), RoutingError> {
+        if !self.chain.is_member() || !self.peer_mgr.is_established() {
+            return Ok(());
         }
 
-        // Check that our section is one of the ones with a minimum length prefix, and if it's not,
-        // forward it to one that is.
-        let min_len_prefix = self.chain().min_len_prefix();
+        self.remove_expired_peers();
 
-        // If we're running in mock-crust mode, and we have relocation interval, don't try to do
-        // section balancing, as it will break things.
-        let forbid_join_balancing = if cfg!(feature = "mock_base") {
-            self.next_relocation_interval.is_some()
-        } else {
-            false
-        };
+        if let Some(prefix) = self.need_resend_expect_candidate_to_prefix() {
+            return self.forward_expect_candidate_to_prefix_section(vote, prefix);
+        }
 
-        if min_len_prefix != *self.our_prefix() && !forbid_join_balancing {
-            let request_content = MessageContent::ExpectCandidate {
-                old_public_id: old_pub_id,
-                old_client_auth: old_client_auth,
-                message_id: message_id,
-            };
-            let src = relocation_dst;
-            let dst = Authority::Section(min_len_prefix.substituted_in(relocation_dst.name()));
-            return self.send_routing_message(src, dst, request_content);
+        if let Some(target_interval) = self.accept_candidate_with_interval(&vote) {
```

Comment:
considering `accept_candidate_with_interval` is a single use method and its just 10lines and this one is also abt 15lines, I'd personally would just inline it, but I'm ok if this layout helps you guys more.

Reply:
This is more in line with single level of abstraction principle, it does an awful lot in this 15 lines.
This allow to highlight all that is important in this function.

Reply:
yeh but making code traversal annoying juggling between multiple methods. RLS helps with that mostly but sure I'd go with what majority are comfortable with.